lucene/sandbox/src/test/org/apache/lucene/search/TestLargeNumHitsTopDocsCollector.java

```diff
@@ -0,0 +1,133 @@
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.RandomIndexWriter;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.LuceneTestCase;
+
+public class TestLargeNumHitsTopDocsCollector extends LuceneTestCase {
+  private Directory dir;
+  private IndexReader reader;
+
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+    dir = newDirectory();
+    RandomIndexWriter writer = new RandomIndexWriter(random(), dir);
+    for (int i = 0; i < 200_000; i++) {
+      writer.addDocument(new Document());
+    }
+    reader = writer.getReader();
+    writer.close();
+  }
+
+  @Override
+  public void tearDown() throws Exception {
+    reader.close();
+    dir.close();
+    dir = null;
+    super.tearDown();
+  }
+  public void testLargeNumAndSparseHits() throws Exception {
+    runNumHits(100_000);
+  }
+
+  public void testSingleNumHit() throws Exception {
+    runNumHits(1);
+  }
+
+  public void testLowNumberOfHits() throws Exception {
+    runNumHits(25);
+  }
+
+  public void testIllegalArguments() throws IOException {
+    Query q = new MatchAllDocsQuery();
+    IndexSearcher searcher = newSearcher(reader);
+    LargeNumHitsTopDocsCollector largeCollector = new LargeNumHitsTopDocsCollector(15);
+    TopScoreDocCollector regularCollector = TopScoreDocCollector.create(15, null, Integer.MAX_VALUE);
+
+    searcher.search(q, largeCollector);
+    searcher.search(q, regularCollector);
+
+    assertEquals(largeCollector.totalHits, regularCollector.totalHits);
+
+    expectThrows(IllegalArgumentException.class, () -> {
+      largeCollector.topDocs(350_000);
+    });
+  }
+
+  public void testNoPQBuild() throws IOException {
+    Query q = new MatchAllDocsQuery();
+    IndexSearcher searcher = newSearcher(reader);
+    LargeNumHitsTopDocsCollector largeCollector = new LargeNumHitsTopDocsCollector(250_000);
+    TopScoreDocCollector regularCollector = TopScoreDocCollector.create(250_000, null, Integer.MAX_VALUE);
+
+    searcher.search(q, largeCollector);
+    searcher.search(q, regularCollector);
+
+    assertEquals(largeCollector.totalHits, regularCollector.totalHits);
+
+    assertEquals(largeCollector.pq, null);
+    assertEquals(largeCollector.pqTop, null);
+  }
+
+  public void testPQBuild() throws IOException {
+    Query q = new MatchAllDocsQuery();
+    IndexSearcher searcher = newSearcher(reader);
+    LargeNumHitsTopDocsCollector largeCollector = new LargeNumHitsTopDocsCollector(100_000);
+    TopScoreDocCollector regularCollector = TopScoreDocCollector.create(100_000, null, Integer.MAX_VALUE);
+
+    searcher.search(q, largeCollector);
+    searcher.search(q, regularCollector);
+
+    assertEquals(largeCollector.totalHits, regularCollector.totalHits);
+
+    assertNotEquals(largeCollector.pq, null);
+    assertNotEquals(largeCollector.pqTop, null);
+  }
+
+  private void runNumHits(int numHits) throws IOException {
+    Query q = new MatchAllDocsQuery();
```

Comment:
Maybe use a query that doesn't produce constant scores to make sure that the collector orders by score? For instance you could add a term to some documents and then build a disjunction here across a query on this term and a MatchAllDocsQuery?

Reply:
+1, updated